bundles/org.openhab.voice.porcupineks/src/main/java/org/openhab/voice/porcupineks/internal/PorcupineKSService.java

```diff
@@ -0,0 +1,329 @@
+        String relativePath;
+        if (path.startsWith("porcupine" + File.separator)) {
+            relativePath = path;
+        } else {
+            relativePath = path.substring(path.lastIndexOf(File.separator + "porcupine" + File.separator) + 1);
+        }
+        File localFile = new File(EXTRACTION_FOLDER,
+                relativePath.substring(relativePath.lastIndexOf(File.separator) + 1));
+        if (!localFile.exists()) {
+            URL porcupineResource = bundleContext.getBundle().getEntry(relativePath);
+            logger.debug("extracting binary {} from bundle to extraction folder", relativePath);
+            extractFromBundle(porcupineResource, localFile);
+        } else {
+            logger.debug("binary {} already extracted", relativePath);
+        }
+        return localFile.toString();
+    }
+
+    private void extractFromBundle(URL resourceUrl, File targetFile) throws IOException {
+        InputStream in = new BufferedInputStream(resourceUrl.openStream());
+        OutputStream out = new BufferedOutputStream(new FileOutputStream(targetFile));
+        byte[] buffer = new byte[1024];
+        int lengthRead;
+        while ((lengthRead = in.read(buffer)) > 0) {
+            out.write(buffer, 0, lengthRead);
+            out.flush();
+        }
+        in.close();
+        out.close();
+    }
+
+    @Override
+    public String getId() {
+        return SERVICE_ID;
+    }
+
+    @Override
+    public String getLabel(@Nullable Locale locale) {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    public Set<Locale> getSupportedLocales() {
+        return Set.of(Locale.ENGLISH, new Locale("es"), Locale.FRENCH, Locale.GERMAN);
+    }
+
+    @Override
+    public Set<AudioFormat> getSupportedFormats() {
+        return Set
+                .of(new AudioFormat(AudioFormat.CONTAINER_WAVE, AudioFormat.CODEC_PCM_SIGNED, false, 16, null, 16000L));
+    }
+
+    @Override
+    public KSServiceHandle spot(KSListener ksListener, AudioStream audioStream, Locale locale, String keyword)
+            throws KSException {
+        Porcupine porcupine;
+        if (config.apiKey.isBlank()) {
+            throw new KSException("Missing pico voice api key");
+        }
+        BundleContext bundleContext = this.bundleContext;
+        if (bundleContext == null) {
+            throw new KSException("Missing bundle context");
+        }
+        try {
+            porcupine = initPorcupine(bundleContext, locale, keyword);
+        } catch (PorcupineException | IOException e) {
+            throw new KSException(e);
+        }
+        Future<?> scheduledTask = executor.submit(() -> processInBackground(porcupine, ksListener, audioStream));
+        return new KSServiceHandle() {
+            @Override
+            public void abort() {
+                logger.debug("stopping service");
+                loop = false;
+                try {
+                    Thread.sleep(0);
```

Comment:
Why addindg a sleep  especially with a 0 value?

Reply:
Sorry for the delay, I woke with fever yesterday and I was resting all day.
I want the background task to end by itself before calling close, in my tests sleeping by 0 is enough for it to complete. Would be better to add a small interval like 100ms?

Reply:
Ok, I understand, that is not a bad idea.
I am not an expert with sleep(0) but just googling, it looks like it is a very special thing.
If you are sure that iit is an appropriate usage, tell me. If not, I would prefer a value > 0, like 100ms.